### PR TITLE
codec_adapter: Add config options to specifify Cadence libraries location

### PR DESCRIPTION
**Pull request by dbaluta**
This will be statically linked with sof when Cadence specific
config options are set.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

### CODE REVIEW

**Review by paulstelian97 — APPROVED** [review]

**Review by iuliana-prodan — APPROVED** [review]

**Review by mrajwa — APPROVED** [review]

**Comment by dbaluta** [comment]
@lgirdwood please do not merge this yet. I will merge it later once Yocto team says this doesn't look horrendously wrong :)

File: src/audio/codec_adapter/CMakeLists.txt
Comment (lgirdwood) [inline src/audio/codec_adapter/CMakeLists.txt:9]: These directories should be set in Kconfig menu. That way Yocto can ship a .config too.

Comment (lgirdwood) [reply src/audio/codec_adapter/CMakeLists.txt:9]: In fact all our static libs should have a $PATH set in Kconfig

Comment (dbaluta) [reply src/audio/codec_adapter/CMakeLists.txt:9]: Is there any standard naming for this? Should be this the path used by all static libs? 

And then we add something like this in CMakeList.txt:

```
 sof_add_static_library(xa_mp3_dec ${STATIC_LIB_PATH}/core_lib/pilibxa_mp3_dec/lib/xa_mp3_dec.a) 
```

Comment (dbaluta) [reply src/audio/codec_adapter/CMakeLists.txt:9]: In fact, I think having the entire binary path set in the Kconfig would make sense. Similar with something like CONFIG_EXTRA_FIRMWARE in the kernel https://cateee.net/lkddb/web-lkddb/EXTRA_FIRMWARE.html

Comment (lgirdwood) [reply src/audio/codec_adapter/CMakeLists.txt:9]: Yeah, this would be a $PATH_CODECX so would be per codec. i.e. on the Kconfig menu when we enable the CODEC then the PATH entry box should activate and allow the user to set a PATH. This also allows Yocto and other embedded distros to carry a config with their own PATHs

**Comment by hongshui3000** [comment]
By the way, sof what should I do if I want to link a private static library. Like this PR, I need to change the Cmake file of sof?

**Comment by dbaluta** [comment]
@hongshui3000 yes.

File: src/audio/codec_adapter/CMakeLists.txt
Comment (dbaluta) [reply src/audio/codec_adapter/CMakeLists.txt]: @lgirdwood OK, sent v2

**Review by lgirdwood — APPROVED** [review]

**Comment by dbaluta** [comment]
@zrombel  you might need to check your internal script for codec_adapter.

**Comment by wszypelt** [comment]
@dbaluta 
It looks like patch is fine, but I'm not sure if the options we use to enable codec support are correct or if something needs to be changed:
echo "CONFIG_COMP_CODEC_ADAPTER=y" >> override.config
echo "CONFIG_CADENCE_CODEC=y" >> override.config
echo "CONFIG_CADENCE_CODEC_WRAPPER=y" >> override.config

**Comment by dbaluta** [comment]
@wszypelt the options are correct. You also need to set this one:

CONFIG_CADENCE_CODEC_WRAPPER_LIB="/path/to/your/cadence/library/"

How do you use the override.config? Is there any options to specify this config at build time?

**Comment by wszypelt** [comment]
@dbaluta 
Thanks for your answer, 
in answer for your questions:
```
cmake -DTOOLCHAIN=xt -DROOT_DIR=$CONFIG_PATH/xtensa-elf -DMEU_PATH=$MEU_PATH -DMEU_PRIVATE_KEY=$PRIVATE_KEY_PATH -DMEU_OPENSSL=$MEU_OPENSSL -DINIT_CONFIG=$make_command  .. 
make $make_command
if [ "$CHROME_BUILD" = "yes" ] ; then make $chrome_make_command; fi
rm -fr override.config
echo "CONFIG_COMP_CODEC_ADAPTER=y" >> override.config
echo "CONFIG_CADENCE_CODEC=y" >> override.config
echo "CONFIG_CADENCE_CODEC_WRAPPER=y" >> override.config
make overrideconfig
make bin -j8 VERBOSE=1 |& tee -a CMakeFiles/stdout.log
```

**Comment by cujomalainey** [comment]
Just random question, why dont we make this more general and have a static lib folder we just link everything in against the firmware rather than on a per lib basis?

**Comment by lgirdwood** [comment]
> Just random question, why dont we make this more general and have a static lib folder we just link everything in against the firmware rather than on a per lib basis?

My thinking behind "per codec" was that if you were developing a closed source codec then you could keep your codec source/git separate from SOF and just point SOF to your build directory for linking. i.e. there is no manual copy stage after building the library to a common SOF library directory and the SOF build will always pick up the latest codec build output.

**Comment by dbaluta** [comment]
@cujomalainey That's a matter of personal taste I guess. Both variants can work very well. I think the current implemented option is a little bit better because we might have codecs from different vendors, each one with their own private repo / directory.

**Comment by lgirdwood** [comment]
@cujomalainey let us know if the directory is a blocker for you ?
@wszypelt @kkarask I assume we would be good to merge once CI path is updated ?

**Comment by cujomalainey** [comment]
By no means a blocker, this is absolutely fine for what I need right now :) was just a comment

**Review by cujomalainey — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kkarask @wszypelt @zrombel good to merge ?

**Comment by zrombel** [comment]
Not yet. I'm setting up QB to match changes. I will leave a comment when changes are ready.

**Comment by zrombel** [comment]
@dbaluta: I have some problems with enabling changes in QB. Qb is applying patch for codec library, can you take a look and tell if any changes need to by added to this patch? To we still need this section in patch ? :
```
+if(CONFIG_COMP_CODEC_ADAPTER)
+  target_link_libraries(sof PRIVATE ${PROJECT_SOURCE_DIR}/xa_dap_vlldp.a)
+endif()
```

[codec_enable.txt](https://github.com/thesofproject/sof/files/6398445/codec_enable.txt)

**Comment by dbaluta** [comment]
@zrombel will have a look.

**Comment by dbaluta** [comment]
@zrombel  you need to update your patch to remove this line:
```
+if(CONFIG_COMP_CODEC_ADAPTER)
+  target_link_libraries(sof PRIVATE ${PROJECT_SOURCE_DIR}/xa_dap_vlldp.a)
+endif()
```

and then set the following config option:

```
CONFIG_CADENCE_CODEC_WRAPPER_LIB="path/to/your//xa_dap_vlldp.a"
```

**Comment by zrombel** [comment]
@lgirdwood @dbaluta QB is ready for this patch so it can be merged. Please inform me 5 min earlier so I can commit changes to QB.

**Comment by lgirdwood** [comment]
@zrombel - merging in 10 mins.

**Comment by zrombel** [comment]
Ok, i'm ready :)

**Comment by dbaluta** [comment]
@zrombel merged!

**Comment by zrombel** [comment]
QB updated, everything is up and running. Thanks!

**Pull request merged**
